src/bit-systems/linear-transform.ts

```diff
@@ -0,0 +1,72 @@
+import { defineQuery, removeComponent } from "bitecs";
+import { Quaternion, Vector3 } from "three";
+import { LinearRotate, LinearScale, LinearTranslate, Object3DTag } from "../bit-components";
+import { HubsWorld } from "../app";
+
+// Working variables
+const _vec3 = new Vector3();
+const _quat = new Quaternion();
+
+const linearTranslateQuery = defineQuery([LinearTranslate, Object3DTag]);
+const linearRotateQuery = defineQuery([LinearRotate, Object3DTag]);
+const linearScaleQuery = defineQuery([LinearScale, Object3DTag]);
+
+export function linearTransformSystem(world: HubsWorld) {
+  linearTranslateQuery(world).forEach(eid => {
+    const duration = LinearTranslate.duration[eid];
+    const destX = LinearTranslate.destinationX[eid];
+    const destY = LinearTranslate.destinationY[eid];
+    const destZ = LinearTranslate.destinationZ[eid];
+
+    const deltaTime = world.time.delta;
+    const obj = world.eid2obj.get(eid)!;
+
+    if (deltaTime >= duration) {
+      obj.position.set(destX, destY, destZ);
+      removeComponent(world, LinearTranslate, eid);
```

Comment:
Feels a bit strange that we are removing the component inside it's own query. Wouldn't it make more sense to remove it in `networked-transform.ts`?

```typescript
      tmpVec.fromArray(NetworkedTransform.position[eid]);
      if (!tmpVec.near(obj.position)) {
        addComponent(world, LinearTranslate, eid);
        LinearTranslate.duration[eid] = millisecondsBetweenTicks;
        LinearTranslate.destinationX[eid] = tmpVec.x;
        LinearTranslate.destinationY[eid] = tmpVec.y;
        LinearTranslate.destinationZ[eid] = tmpVec.z;
      } else {
        removeComponent(world, LinearTranslate, eid);
      }
```

and let this system do just what is responsible for, interpolating:

```typescript
linearTranslateQuery(world).forEach(eid => {
    const duration = LinearTranslate.duration[eid];
    const destX = LinearTranslate.destinationX[eid];
    const destY = LinearTranslate.destinationY[eid];
    const destZ = LinearTranslate.destinationZ[eid];

    const deltaTime = world.time.delta;
    const obj = world.eid2obj.get(eid)!;

    obj.position.lerp(_vec3.set(destX, destY, destZ), deltaTime / duration);
    LinearTranslate.duration[eid] = duration - deltaTime;
    obj.matrixNeedsUpdate = true;
  });
```

Reply:
I don't think it is so strange to remove a component inside a query especially if it does operations across frames and finishes at specified time or event. MediaLoader does it, too.

And LinearTranslate/Rotate/Scale and their system are designed as common components and systems that linearly (against time) update object's transform to target values using the specified duration time. I think removing component in that system is natural and makes them decoupled from other components and systems.

But if removing a component in its query can cause a bug in certain scenarios, we may need to revisit and write it in the internal design document. Would you please share if you have?